File: facade_test.go
package yaylib

import (
	"context"
	"testing"
)

// PORTING:S33
// Flat surface: in Go every operation is promoted onto *Client via
// embed, and a hand-written wrapper shadows the embedded generated op
// of the same name. Both are guaranteed structurally — referencing the
// promoted method compiles only if promotion holds, and LoginWithEmail
// resolves to the cached wrapper (its builder type), not the generated
// request. Hermetic: no network is performed.
func TestFlatFacade_OpsPromotedAndWrapperShadows(t *testing.T) {
	c := NewClient()
	defer c.Close()

	// Taking these as method values compiles only if the operations
	// are promoted directly onto *Client (no service prefix) — the
	// embed-promotion that gives Go the flat surface for free.
	_ = c.GetTimeline
	_ = c.GetRecommendedTimeline

	// LoginWithEmail resolves to the cached wrapper (returns its
	// chain builder), shadowing the embedded generated op of the same
	// name — the embed-shadowing rule. Reaching .Email/.Password on
	// the result compiles only against the wrapper builder.
	_ = c.LoginWithEmail(context.Background()).Email("").Password("")

	if c.UserID != 0 {
		t.Fatalf("fresh client should not be logged in, got UserID=%d", c.UserID)
	}
}

File: packages/typescript/tests/facade.test.ts
// Flat operation facade (PORTING.md §15 S33 / §2).
//
// Every operation is reachable as client.<op> with no service prefix,
// and a hand-written wrapper of the same name shadows the generated op.
// Pure local check — no server is involved.

import { Client } from "../src/index";

let failed = 0;
function assert(name: string, cond: boolean, detail = ""): void {
  if (cond) console.log(`PASS ${name}`);
  else {
    failed++;
    console.log(`FAIL ${name}${detail ? `\n  ${detail}` : ""}`);
  }
}

// PORTING:S33
async function flatOpsAndWrapperShadow(): Promise<void> {
  const c = new Client();

  // Operations are promoted onto the client directly (no client.postsAPI.).
  assert(
    "client.getTimeline is a flat method",
    typeof (c as unknown as Record<string, unknown>).getTimeline === "function",
  );
  assert(
    "client.getRecommendedTimeline is a flat method",
    typeof (c as unknown as Record<string, unknown>).getRecommendedTimeline ===
      "function",
  );

  // The hand-written cached wrapper shadows the generated loginWithEmail
  // op: empty credentials fail fast with the wrapper's validation error,
  // never an HTTP attempt.
  let shadow = false;
  try {
    await c.loginWithEmail({ email: "", password: "" });
  } catch (e) {
    shadow = /requires email and password/.test(String((e as Error)?.message));
  }
  assert("loginWithEmail resolves to the cached wrapper (shadow)", shadow);
}

(async () => {
  await flatOpsAndWrapperShadow();
  process.exit(failed === 0 ? 0 : 1);
})().catch((e) => {
  console.error("test crashed:", e);
  process.exit(2);
});
